### PR TITLE
Add docs on clever errors

### PR DESCRIPTION
**Pull request by tzakian**
Adds docs on clever errors.

### CODE REVIEW

File: reference/src/abort-and-assert/clever-errors.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+## Clever Errors and Macros
+
+The line number information in clever abort codes are derived from the source file at the location
+where the abort occurs. In particular, for a function this will be the line number within in the
+function, however for macros, this will be the location where the macro is invoked. This can be
+quite useful when writing macros as it provides a way for users to use macros that may raise abort
+conditions and still get useful error messages.
+
+```move
+module 0x42::macro_exporter {
```
Comment (damirka) [inline reference/src/abort-and-assert/clever-errors.md:126]: For coherence, I would split it into two codeblocks and use module labels.

File: reference/src/abort-and-assert/clever-errors.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+this, post-processing is required to go from the `u64` abort code value to a human-readable error
+message. This post-processing is automatically performed by the Sui GraphQL server, as well as the
+Sui CLI. If you want to manually decode a clever abort code, you can use the process outlined in
+[Inflating Clever Abort Codes](#inflating-clever-abort-codes) to do so.
+
+> Clever errors include source line information amongst other data. Because of this their value may
+> change due to any changes in the source file (e.g., due to auto-formatting, adding a new module
+> member, or adding a newline).
+
+## Clever Abort Codes
```
Comment (damirka) [inline reference/src/abort-and-assert/clever-errors.md:16]: Can you add a link to a matching page in the `book/`?

File: reference/src/abort-and-assert/clever-errors.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
+# Clever Errors
+
+Clever errors are a feature that allows for more informative error messages when an assertion fails
+or an abort is raised. They are a source feature and compile to a `u64` abort code value that
+contains the information needed to access the line number, constant name, and constant value given
+the clever error code and the module that the clever error constant was declared in. Because of
+this, post-processing is required to go from the `u64` abort code value to a human-readable error
```
Comment (tnowacki) [inline reference/src/abort-and-assert/clever-errors.md:7]:
```suggestion
this compilation, post-processing is required to go from the `u64` abort code value to a human-readable error
```

File: reference/src/abort-and-assert/clever-errors.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+message. This post-processing is automatically performed by the Sui GraphQL server, as well as the
```
Comment (tnowacki) [inline reference/src/abort-and-assert/clever-errors.md:8]:
```suggestion
message. The post-processing is automatically performed by the Sui GraphQL server, as well as the
```

Comment (tnowacki) [reply reference/src/abort-and-assert/clever-errors.md:8]: I think this just reads cleaner given the previous sentence.

File: reference/src/abort-and-assert/clever-errors.md (hunk 3 above)
Comment (tnowacki) [reply reference/src/abort-and-assert/clever-errors.md:7]: Trying to be less ambiguous.

File: reference/src/abort-and-assert/clever-errors.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+The exact formatting of this message may vary depending on the tooling used to decode the clever
+error however all of the information needed to generate a human-readable error message like the
+above is present in the `u64` abort code when coupled with the module where the error occurred.
+
+> Clever abort code values do _not_ need to be a `vector<u8>` -- it can be any valid constant type
+> in Move.
+
+## Assertions with no Abort Codes
+
+Assertions, and `abort` statements without an abort code will automatically derive an abort code
```
Comment (tnowacki) [inline reference/src/abort-and-assert/clever-errors.md:77]:
```suggestion
Assertions and `abort` statements without an abort code will automatically derive an abort code
```

File: reference/src/abort-and-assert/clever-errors.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+1. A set tag-bit that indicates that the abort code is a clever abort code.
+2. The line number of where the abort occured in the source file (e.g., 7).
+3. The index in the module's identifier table for the constant's name (e.g., `EIsThree`).
+4. The index of the constant's value in the module's constant table (e.g., `b"The value is three"`).
+
+In hex, if `double_except_three(3)` is called, it will abort with a `u64` abort code as follows:
+
+```
+0x8000_7000_1000_0000
```
Comment (tnowacki) [inline reference/src/abort-and-assert/clever-errors.md:52]:
```suggestion
0x8000_0007_0001_0000
```
Is this correct? I was surprised that the value was set into the higher bits there.

File: reference/src/abort-and-assert/clever-errors.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+2. The line number of where the abort occured in the source file (e.g., 6).
+3. A sentinel value of `0xffff` for the index into the module's identifier table for the constant's
+   name.
+4. A sentinel value of `0xffff` for the index of the constant's value in the module's constant
+   table.
+
+In hex, if `assert_false(3)` is called, it will abort with a `u64` abort code as follows:
+
+```
+0x8000_4000_ffff_ffff
```
Comment (tnowacki) [inline reference/src/abort-and-assert/clever-errors.md:107]: Let's double check this with the one above

File: reference/src/abort-and-assert/clever-errors.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    if identifier_index == 0xffff && constant_index == 0xffff {
+        return;
+    }
+
+    // Only needed if constant name and value are not 0xffff
+    let module: CompiledModule = fetch_module(package_id, module_name);
+
+    // Print the constant name (if any)
+    if identifier_index != 0xffff {
+        let constant_name = get_identifier_at_table_index(module, identifier_index);
```
Comment (tnowacki) [inline reference/src/abort-and-assert/clever-errors.md:218]: > `get_identifier_at_table_index`

I appreciate this is pseudo code, but do we have a method for this?

File: reference/src/abort-and-assert/clever-errors.md (hunk 6 above)
Comment (tzakian) [reply reference/src/abort-and-assert/clever-errors.md:52]: Good catch! I indeed got the order swapped here.

File: reference/src/abort-and-assert/clever-errors.md (hunk 8 above)
Comment (tzakian) [reply reference/src/abort-and-assert/clever-errors.md:218]: We do, but its name is not quite as descriptive as I would like here (`identifier_at`). I'll try and make it a bit closer, but I think having a more descriptive name here will be beneficial.

File: reference/src/abort-and-assert/clever-errors.md (hunk 2 above)
Comment (tzakian) [reply reference/src/abort-and-assert/clever-errors.md:16]: We don't have one in the `book/` (at least not yet) so not really anything to reference here unless I'm misunderstanding something.

**Review by tnowacki — APPROVED** [review]

**Pull request merged**